### ISSUE TITLE
Support User-Agent Client Hints override

### ISSUE THREAD

**Issue by rowan-m** [puppeteer/puppeteer#7269]
[`page.setUserAgent(userAgent)`](https://pptr.dev/#?product=Puppeteer&version=v9.1.1&show=api-pagesetuseragentuseragent) allows Puppeteer to override the user-agent. This passes through to [`Network.setUserAgentOverride(userAgent)`](https://chromedevtools.github.io/devtools-protocol/tot/Network/#method-setUserAgentOverride) on the DevTools protocol side.

It looks like this does support (currently as an experimental parameter) sending [`userAgentMetadata`](https://chromedevtools.github.io/devtools-protocol/tot/Emulation/#type-UserAgentMetadata) as an optional parameter.

On the site side, we're [encouraging developers to feature detect support for User-Agent Client Hints](https://web.dev/migrate-to-ua-ch/) and then only fallback to the user-agent string when not available. As such, if a site follows this best practice then it will read the unaltered data from User-Agent Client Hints as opposed to whatever has been set in the string.

I think I'd suggest something like `page.setUserAgentData(userAgentData)` here that accepts a parameter in the same format as the DevTools protocol.

**Comment by rowan-m** [comment]
Oh, and as this initially looks like a relatively simple addition, I can look at putting in a PR if the suggested change makes sense.

**Comment by mathiasbynens** [comment]
Another option would be to extend the current API with an optional parameter for the additional data:

```js
page.setUserAgent(userAgentString)
// →
page.setUserAgent(userAgentString, additionalUserAgentData)
```

I think I weakly prefer that version, as it seems folks who want to override UA Client Hints would also want to override the UA string, and it'd be nice if they didn’t need two separate API calls for that. WDYT?

>  I can look at putting in a PR if the suggested change makes sense.

Please go ahead :)

**Comment by rowan-m** [comment]
I could perhaps… do both? This encourages setting `userAgentData` along with the string to avoid getting out of sync, but also provides a clean interface for people who don't need to worry about the string and just want to set the hints.

**Comment by mathiasbynens** [comment]
Hmm… My gut feeling is that it’s a niche use case to only want to set the UA-related client hints without also changing the UA string. Through that lens, `page.setUserAgent(userAgentString, additionalUserAgentData)` seems sufficient. To support your use case, we could allow `userAgentString = null` to signal “don’t override the `User-Agent` header”, which would be consistent with [`page.emulateMediaType(type)`](https://github.com/puppeteer/puppeteer/blob/main/docs/api.md#pageemulatemediatypetype).

The “extend `page.setUserAgent`” approach:

```js
// 1. Set only the `User-Agent` header (and `navigator.userAgent`).
page.setUserAgent(userAgentString);
// 2. Set the `User-Agent` header + the UA-related Client Hints.
page.setUserAgent(userAgentString, additionalUserAgentData);
// 3. Set only the UA-related Client Hints (niche use case).
page.setUserAgent(null, additionalUserAgentData);
```

The “do both” approach:

```js
// 1. Set only the `User-Agent` header (and `navigator.userAgent`).
page.setUserAgent(userAgentString);
// 2. Set the `User-Agent` header + the UA-related Client Hints.
page.setUserAgent(userAgentString, additionalUserAgentData);
// 3. Set only the UA-related Client Hints (niche use case).
page.setUserAgentData(additionalUserAgentData);
```

The only difference is in `3`, which I feel is an edge case. Based on this assumption (which might be wrong!) I'd rather not add a new API just for this case.

`page.setUserAgent` already affects the `User-Agent` HTTP header and `navigator.userAgent`. Now that there's more UA-related data that should be configurable (UA Client Hints), it makes sense for the same API to deal with it, IMHO. (I tend to be conservative about adding new APIs because it’s painful to deprecate/remove them later, and in this case it seems we can get away without a new API.)

WDYT?

**Comment by rowan-m** [comment]
I believe this is now closed with the merging of #7378

**Comment by mathiasbynens** [comment]
Hurray! Thanks for implementing this, @rowan-m!
